server/src/main/java/com/flipkart/varadhi/CoreServices.java

```diff
@@ -81,9 +81,7 @@ private <T> T loadClass(String className) {
             }
             throw new InvalidConfigException("No class provided.");
         } catch (Exception e) {
-            String errorMsg = String.format("Fail to load class %s.", className);
-            log.error(errorMsg, e);
-            throw new InvalidConfigException(e);
+            throw new InvalidConfigException(String.format("Fail to load class %s.", className), e);
```

Comment:
Any thoughts for removing logging ?

Reply:
logging error and then throwing the same leads to redundant logs.

For eg: 
### Before
```logs
2023-06-26 12:19:05 ERROR CoreServices:85 - Fail to load class com.flipkart.varadhi.pulsar.PulsarStackProvider.
java.lang.ClassNotFoundException: com.flipkart.varadhi.pulsar.PulsarStackProvider
        at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
        at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Class.java:375) ~[?:?]
        at com.flipkart.varadhi.CoreServices.loadClass(CoreServices.java:79) [main/:?]
        at com.flipkart.varadhi.CoreServices.setupMessagingStackProvider(CoreServices.java:71) [main/:?]
        at com.flipkart.varadhi.CoreServices.<init>(CoreServices.java:41) [main/:?]
        at com.flipkart.varadhi.Server.main(Server.java:22) [main/:?]
2023-06-26 12:19:05 ERROR Server:27 - Failed to initialise the server.
com.flipkart.varadhi.exceptions.InvalidConfigException: com.flipkart.varadhi.pulsar.PulsarStackProvider
        at com.flipkart.varadhi.CoreServices.loadClass(CoreServices.java:86) ~[main/:?]
        at com.flipkart.varadhi.CoreServices.setupMessagingStackProvider(CoreServices.java:71) ~[main/:?]
        at com.flipkart.varadhi.CoreServices.<init>(CoreServices.java:41) ~[main/:?]
        at com.flipkart.varadhi.Server.main(Server.java:22) [main/:?]
Caused by: java.lang.ClassNotFoundException: com.flipkart.varadhi.pulsar.PulsarStackProvider
        at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
        at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Class.java:375) ~[?:?]
        at com.flipkart.varadhi.CoreServices.loadClass(CoreServices.java:79) ~[main/:?]
        ... 3 more
Failed to initialise the server:com.flipkart.varadhi.exceptions.InvalidConfigException: com.flipkart.varadhi.pulsar.PulsarStackProvider
```

### After
```logs
2023-06-26 12:19:53 ERROR Server:27 - Failed to initialise the server.
com.flipkart.varadhi.exceptions.InvalidConfigException: Fail to load class com.flipkart.varadhi.pulsar.PulsarStackProvider.
        at com.flipkart.varadhi.CoreServices.loadClass(CoreServices.java:84) ~[main/:?]
        at com.flipkart.varadhi.CoreServices.setupMessagingStackProvider(CoreServices.java:71) ~[main/:?]
        at com.flipkart.varadhi.CoreServices.<init>(CoreServices.java:41) ~[main/:?]
        at com.flipkart.varadhi.Server.main(Server.java:22) [main/:?]
Caused by: java.lang.ClassNotFoundException: com.flipkart.varadhi.pulsar.PulsarStackProvider
        at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641) ~[?:?]
        at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188) ~[?:?]
        at java.lang.ClassLoader.loadClass(ClassLoader.java:520) ~[?:?]
        at java.lang.Class.forName0(Native Method) ~[?:?]
        at java.lang.Class.forName(Class.java:375) ~[?:?]
        at com.flipkart.varadhi.CoreServices.loadClass(CoreServices.java:79) ~[main/:?]
        ... 3 more
```

Reply:
I try to follow the rule for error logs which says: Don't log if you are throwing. can log if you are handling.

Reply:
generally I tend to follow log at source of the event to capture all data along that leads to exception.  
But I guess, we can avoid exception logging from that and keep only data part and leave exception logging to either default exception handler at base or log where it is being handled ?

Reply:
We can still capture what we want to capture in the "exception" that we are throwing.